### PR TITLE
[DDW-595] Warn user when pasting an address of the same wallet

### PR DESCRIPTION
**Pull request by aleksandardjordjeviciohk**
This PR adds warning message as tooltip to the user when pasting an address of the same wallet in the send form.

## POTENTIALLY BREAKING CHANGES :boom:

This PR includes an improved Input component in react-polymorph to make it more configurable and support use cases like in this PR: https://github.com/input-output-hk/react-polymorph/pull/173

**Input and FormField components from now on support:**

- Showing form field error on hover
- Showing form field error on focus
- Styling the form field errors differently via injected css variables

The breaking change is that ALL form fields in the whole application now show their errors on focus/hover by default - this needs to be checked carefully before merging this PR.

## TODO ✅ 

- [x] Update copy for tooltip + translations

## Screenshots

Screen Shot 2021-04-04 at 11 38 13 PM

Screen Shot 2021-04-04 at 11 38 35 PM

---

## Testing Checklist 👀 

- [Slack QA thread](https://input-output-rnd.slack.com/archives/GGKFXSKC6/p1617607697323300)

- [ ] ❗we need to test out all form fields across the whole app, so we could be sure that nothing is broken.

---
## Test Scenarios

**Scenario 1 - Use Address from Wallet A on Wallet A**
```gherkin
Given that Daedalus is synced
And I have synced at least one Shelley wallet with funds
And I am on the "Send" screen from Wallet A
When I paste an address from Wallet A in the "Receiver" input
And I add a valid amount on the "Ada" input
Then I can see the "Send" button is enabled
And the "Receiver" input is outlined in orange
And the following alert is displayed on a tooltip

ENGLISH: This address belongs to your [Wallet name] wallet. No assets will be transferred, but fee will be charged.
JAPANESE: このアドレスはあなたのものです[Wallet name]財布。資産は譲渡されませんが、手数料がかかります
````
**Scenario 2 - Use Address from Wallet B on Wallet A**
```gherkin
Given that Daedalus is synced
And I have synced at least two Shelley wallet with funds
And I am on the "Send" screen from Wallet A
When I paste an address from Wallet B in the "Receiver" input
And I add a valid amount on the "Ada" input
Then I can see the "Send" button is enabled
And the "Receiver" input shows no alerts
```

---

### Basics

- [ ] PR has been assigned and has appropriate labels (`feature`/`bug`/`chore`, `release-x.x.x`)
- [ ] PR is updated to the most recent version of the target branch (and there are no conflicts)
- [ ] PR has a good description that summarizes all changes
- [ ] PR has default-sized Daedalus window screenshots or animated GIFs of important UI changes:
  - [ ] In English
  - [ ] In Japanese
- [ ] CHANGELOG entry has been added to the top of the appropriate section (*Features*, *Fixes*, *Chores*) and is linked to the correct PR on GitHub
- [ ] Automated tests: All acceptance and unit tests are passing (`yarn test`)
- [ ] Manual tests (minimum tests should cover newly added feature/fix): App works correctly in *development* build (`yarn dev`)
- [ ] Manual tests (minimum tests should cover newly added feature/fix): App works correctly in *production* build (`yarn package` / CI builds)
- [ ] There are no *flow* errors or warnings (`yarn flow:test`)
- [ ] There are no *lint* errors or warnings (`yarn lint`)
- [ ] There are no *prettier* errors or warnings (`yarn prettier:check`)
- [ ] There are no missing translations (running `yarn manage:translations` produces no changes)
- [ ] Text changes are proofread and approved (Jane Wild / Amy Reeve)
- [ ] Japanese text changes are proofread and approved (Junko Oda)
- [ ] UI changes look good in all themes (Alexander Rukin)
- [ ] Storybook works and no stories are broken (`yarn storybook`)
- [ ] In case of dependency changes `yarn.lock` file is updated

### Code Quality
- [ ] Important parts of the code are properly commented and documented
- [ ] Code is properly typed with flow
- [ ] React components are split-up enough to avoid unnecessary re-renderings
- [ ] Any code that only works in main process is neatly separated from components

### Testing
- [ ] New feature/change is covered by acceptance tests
- [ ] New feature/change is manually tested and approved by QA team
- [ ] All existing acceptance tests are still up-to-date
- [ ] New feature/change is covered by Daedalus Testing scenario
- [ ] All existing Daedalus Testing scenarios are still up-to-date

### After Review
- [ ] Merge the PR
- [ ] Delete the source branch
- [ ] Move the ticket to `done` column on the YouTrack board
- [ ] Update Slack QA thread by marking it with a green checkmark

### CODE REVIEW

File: source/renderer/app/components/wallet/WalletSendForm.js [hunk 1]
```diff
@@ -696,28 +703,64 @@ export default class WalletSendForm extends Component<Props, State> {
...
+              error={receiverField.error}
+              onChange={(value) => {
+                receiverField.onChange(value || '');
+                this.setState({
+                  isResetButtonDisabled: false,
+                });
+              }}
+              onKeyPress={this.handleSubmitOnEnter}
+            />
+          )}
```
Comment (DominikGuzei) [inline source/renderer/app/components/wallet/WalletSendForm.js:763]: @aleksandardjordjeviciohk let's refactor the duplication of the `Input` component here - we could either save it into a variable above the return and re-use it for the popover condition logic or maybe even better: use a conditional [disabled](https://github.com/atomiks/tippyjs-react#disabled-boolean) boolean flag on the `PopOver` (haven't tried that yet but would be good to test if that works)

**Review by DominikGuzei — CHANGES_REQUESTED** [review]
**There is a UX issue:**
The warning briefly shows up but disappears after a few seconds (I guess this has something to do with the wallets data refresh every 5 seconds - maybe this is overriding the boolean flag again?).

The length of how long it is shown varies (again, probably because of the wallets data refresh interval kicking in at different times)

https://user-images.githubusercontent.com/172414/113559819-53810f00-9602-11eb-81c4-650b0f5e4839.mp4

**Comment by gabriela-ponce** [comment]
@aleksandardjordjeviciohk I agree with @DominikGuzei. The outline on the field sometimes is displayed for a split second and you can't see the tooltip because it hides very quickly. Maybe it can be displayed all the time as the red outline when the address is incorrect? This way, every time the user hovers the field, they should be able to see the tooltip. Check [this video](https://drive.google.com/file/d/1rdXhsjQr4UGUtiFIWb5M5kdx9K2sszdP/view?usp=sharing).

**Comment by gabriela-ponce** [comment]
@aleksandardjordjeviciohk The previous issue is fixed, nice work 👍 

I found other issues in this instance: 

- When you paste an address from the same wallet and then change the value to make it not valid, the yellow outline and tooltip are still displayed. Check [this video](https://drive.google.com/file/d/1R8Aqpk-SFg3auNV5LQ4s32PjHqwmmvIa/view?usp=sharing).

- Also, if you clear the field by clicking on "x", you'll see the same message after inserting any value. Check [this video.](https://drive.google.com/file/d/1RiZ3V-ROjx1kD0QSpnFBYQOqDu1hIqrc/view?usp=sharing)

File: source/renderer/app/components/wallet/WalletSendForm.js (hunk 1 above)
Comment (aleksandardjordjeviciohk) [reply source/renderer/app/components/wallet/WalletSendForm.js]: Fixed

**Comment by aleksandardjordjeviciohk** [comment]
> @aleksandardjordjeviciohk The previous issue is fixed, nice work 👍
> 
> I found other issues in this instance:
> 
> * When you paste an address from the same wallet and then change the value to make it not valid, the yellow outline and tooltip are still displayed. Check [this video](https://drive.google.com/file/d/1R8Aqpk-SFg3auNV5LQ4s32PjHqwmmvIa/view?usp=sharing).
> * Also, if you clear the field by clicking on "x", you'll see the same message after inserting any value. Check [this video.](https://drive.google.com/file/d/1RiZ3V-ROjx1kD0QSpnFBYQOqDu1hIqrc/view?usp=sharing)

These issues are fixed, pls recheck @gabriela-ponce

File: source/renderer/app/components/wallet/WalletSendForm.js (hunk 1 above)
Comment (aleksandardjordjeviciohk) [reply source/renderer/app/components/wallet/WalletSendForm.js]: Refactored @DominikGuzei

**Comment by gabriela-ponce** [comment]
@aleksandardjordjeviciohk The fixes look good on the build 17524. 
There is a small issue now with the "please insert a valid address" tooltip. In production, this tooltip is displayed [if you're writing on the input or when it is in focus](https://drive.google.com/file/d/1PvFJRFX6G9-X6_GPIcDZ20iWb0YOEXHy/view?usp=sharing). On build 17524, you can't see the tooltip until you click on the field (even if it's in focus). Check videos [1](https://drive.google.com/file/d/1G7o7T3gfOwL1Xi5-G2YqWXH6tV9xpvxq/view?usp=sharing) and [2](https://drive.google.com/file/d/1-Sc8srGpTXhIB7TQLgNPdQEranymCtqA/view?usp=sharing).

**Comment by aleksandardjordjeviciohk** [comment]
> @aleksandardjordjeviciohk The fixes look good on the build 17524.
> There is a small issue now with the "please insert a valid address" tooltip. In production, this tooltip is displayed [if you're writing on the input or when it is in focus](https://drive.google.com/file/d/1PvFJRFX6G9-X6_GPIcDZ20iWb0YOEXHy/view?usp=sharing). On build 17524, you can't see the tooltip until you click on the field (even if it's in focus). Check videos [1](https://drive.google.com/file/d/1G7o7T3gfOwL1Xi5-G2YqWXH6tV9xpvxq/view?usp=sharing) and [2](https://drive.google.com/file/d/1-Sc8srGpTXhIB7TQLgNPdQEranymCtqA/view?usp=sharing).

@gabriela-ponce @DominikGuzei can you guys review again

**Comment by DominikGuzei** [comment]
@aleksandardjordjeviciohk the yellow warning does not show (only the borders are yellow) when pasting in an address of the same wallet:

https://user-images.githubusercontent.com/172414/114077369-02924480-98a8-11eb-93be-c518dfaff210.mp4

I guess that's caused by the instant switch to the ADA field … but I would expect to see the warning test immediately and not switch to the ADA field right away (since the user probably doesn't want to finish this operation anyway)

**Comment by gabriela-ponce** [comment]
@aleksandardjordjeviciohk The issue on [this](https://github.com/input-output-hk/daedalus/pull/2506#issuecomment-814940341) comment is fixed, nice job 👍 

> There is a small issue now with the "please insert a valid address" tooltip. In production, this tooltip is displayed if you're writing on the input or when it is in focus. On build 17524, you can't see the tooltip until you click on the field (even if it's in focus). Check videos 1 and 2.

I could replicate the issue @DominikGuzei mentioned in [the previous comment](https://github.com/input-output-hk/daedalus/pull/2506#issuecomment-816040011). I agree that the tooltip should be displayed in that scenario.

**Comment by aleksandardjordjeviciohk** [comment]
> @aleksandardjordjeviciohk the yellow warning does not show (only the borders are yellow) when pasting in an address of the same wallet:
> 
>  CleanShot.2021-04-08.at.20.19.40.mp4 
> I guess that's caused by the instant switch to the ADA field … but I would expect to see the warning test immediately and not switch to the ADA field right away (since the user probably doesn't want to finish this operation anyway)

fixed @DominikGuzei @gabriela-ponce

File: source/renderer/app/components/wallet/WalletSendForm.scss [hunk 2]
```diff
@@ -317,3 +323,6 @@
     }
   }
 }
+
+.sameWalletTooltipContent {
+}
```
Comment (DominikGuzei) [inline source/renderer/app/components/wallet/WalletSendForm.scss:328]: I guess this can be removed

**Review by DominikGuzei — CHANGES_REQUESTED** [review]
Works much better now 🎉 the only issue i found is that after you change focus to the ADA amount input the yellow border disappears from he receiver field:


https://user-images.githubusercontent.com/172414/114407771-0a543080-9ba9-11eb-8315-0dce0d67f1db.mp4

File: source/renderer/app/components/wallet/WalletSendForm.js [hunk 3]
```diff
@@ -299,8 +301,16 @@ export default class WalletSendForm extends Component<Props, State> {
...
-    return get(asset, 'value', false);
+  isAddressFromSameWallet = () => {
+    const { addressFromSameWallet } = this.props;
+    const receiverField = this.form.$('receiver');
+    return (
+      this.hasReceiverValue() &&
+      addressFromSameWallet &&
+      receiverField.focused &&
+      receiverField.isValid &&
+      !receiverField.error
```
[the remark this answers is not in the document]
Comment (aleksandardjordjeviciohk) [reply source/renderer/app/components/wallet/WalletSendForm.js]: @DominikGuzei fixed

File: source/renderer/app/components/wallet/WalletSendForm.scss (hunk 2 above)
Comment (aleksandardjordjeviciohk) [reply source/renderer/app/components/wallet/WalletSendForm.scss]: @DominikGuzei fixed

**Comment by aleksandardjordjeviciohk** [comment]
> Works much better now 🎉 the only issue i found is that after you change focus to the ADA amount input the yellow border disappears from he receiver field:
> 
>  CleanShot.2021-04-12.at.15.55.58.mp4

Fixed

**Comment by gabriela-ponce** [comment]
@aleksandardjordjeviciohk The issue @DominikGuzei mentioned is fixed. However, now the tooltip is always displayed, and it should hide when the user switched the focus and be displayed again on hover. Check [this screenshot](https://drive.google.com/file/d/14fo50yvN1kR52zZe5Jq5n_nNW5zKB6no/view?usp=sharing).
But, when the user inserts the address for the first time, as it is valid, the focus change to the next field, so not sure how should we handle this. Maybe show the tooltip until the user starts to type on the next field?
The outline should always be displayed in my opinion.

**Comment by aleksandardjordjeviciohk** [comment]
> @aleksandardjordjeviciohk The issue @DominikGuzei mentioned is fixed. However, now the tooltip is always displayed, and it should hide when the user switched the focus and be displayed again on hover. Check [this screenshot](https://drive.google.com/file/d/14fo50yvN1kR52zZe5Jq5n_nNW5zKB6no/view?usp=sharing).
> But, when the user inserts the address for the first time, as it is valid, the focus change to the next field, so not sure how should we handle this. Maybe show the tooltip until the user starts to type on the next field?
> The outline should always be displayed in my opinion.

@DominikGuzei @nikolaglumac any thoughts?

**Comment by nikolaglumac** [comment]
Let's simply not change focus automatically in that case?

**Comment by alexander-rukin** [comment]
This box seems to be shifted to the bottom more (1px) - [screenshot](https://user-images.githubusercontent.com/2539345/114834992-cbf18800-9dd9-11eb-9e11-ae2be9b5ddbd.png)

**Comment by alexander-rukin** [comment]
I'm pretty sure that this whole story with searchable dropdowns was created to be able search [this](https://user-images.githubusercontent.com/2539345/114835812-a3b65900-9dda-11eb-85a4-94f2a4a459b0.png)
Are we going to implement this?

**Comment by nikolaglumac** [comment]
> I'm pretty sure that this whole story with searchable dropdowns was created to be able search [this](https://user-images.githubusercontent.com/2539345/114835812-a3b65900-9dda-11eb-85a4-94f2a4a459b0.png)
> Are we going to implement this?

@alexander-rukin I believe you have posted this comment on a wrong PR...

File: CHANGELOG.md [hunk 4]
```diff
@@ -14,6 +14,10 @@ Changelog
 - Fixed fee calculation edge cases in wallet send form ([PR 2501](https://github.com/input-output-hk/daedalus/pull/2501))
 - Handle empty strings in transaction metadata correctly ([PR 2503](https://github.com/input-output-hk/daedalus/pull/2503))
 
+### Features
+
+- Added warning message when pasting an address of the same wallet in the send form ([PR 2506](https://github.com/input-output-hk/daedalus/pull/2506))
```
Comment (nikolaglumac) [inline CHANGELOG.md:19]: @aleksandardjordjeviciohk please move under "vNext"

**Comment by alexander-rukin** [comment]
> @alexander-rukin I believe you have posted this comment on a wrong PR...

thanks

File: CHANGELOG.md (hunk 4 above)
Comment (aleksandardjordjeviciohk) [reply CHANGELOG.md:25]: Fixed

**Comment by aleksandardjordjeviciohk** [comment]
> Let's simply not change focus automatically in that case?

Fixed

**Review by gabriela-ponce — APPROVED** [review]
Looks good 👍

File: source/renderer/app/components/wallet/WalletSendForm.js [hunk 5]
```diff
@@ -67,6 +67,8 @@ type Props = {
   onOpenDialogAction: Function,
   onUnsetActiveAssetFingerprint: Function,
   onExternalLinkClick: Function,
+  addressFromSameWallet: boolean,
```
Comment (nikolaglumac) [inline source/renderer/app/components/wallet/WalletSendForm.js:70]: @aleksandardjordjeviciohk please rename to `isAddressFromSameWallet`

File: source/renderer/app/components/wallet/WalletSendForm.js [hunk 6]
```diff
@@ -299,8 +301,15 @@ export default class WalletSendForm extends Component<Props, State> {
     return receiverField.value.length > 0;
   };
 
-  hasAssetValue = (asset: Field) => {
```
Comment (nikolaglumac) [inline source/renderer/app/components/wallet/WalletSendForm.js:302]: @aleksandardjordjeviciohk what happened with this? Why was it removed?

**Review by nikolaglumac — CHANGES_REQUESTED** [review]

**Comment by alexander-rukin** [comment]
I think that even focusing other field we still need to show popup on hover 
image

File: source/renderer/app/components/wallet/WalletSendForm.js (hunk 6 above)
Comment (aleksandardjordjeviciohk) [reply source/renderer/app/components/wallet/WalletSendForm.js:302]: Not being used at this component, obsolete method.

File: source/renderer/app/components/wallet/WalletSendForm.js (hunk 5 above)
Comment (aleksandardjordjeviciohk) [reply source/renderer/app/components/wallet/WalletSendForm.js:70]: Fixed

**Comment by aleksandardjordjeviciohk** [comment]
> I think that even focusing other field we still need to show popup on hover
> image

Makes sense, i added show on hover, you can recheck it @alexander-rukin

**Comment by DominikGuzei** [comment]
I just discussed with @aleksandardjordjeviciohk that the `Input / Formfield` components should fully support these use cases required for this feature. So added a WIP flag and will make the necessary changes in react-polymorph and update this PR then.

**Comment by DominikGuzei** [comment]
@aleksandardjordjeviciohk @gabriela-ponce @nikolaglumac i improved the `Input` component in react-polymorph to make it more configurable and support use cases like in this PR: https://github.com/input-output-hk/react-polymorph/pull/173

I also updated this PR and fixed any remaining issues with tab indexes etc. - IMO it looks good now and works solid 👍

**Review by DominikGuzei — APPROVED** [review]

**Review by nikolaglumac — CHANGES_REQUESTED** [review]
@aleksandardjordjeviciohk please also prevent autofocus of the token amount field in case user enters the same address (for this you need to click on "SEND" in one of the token boxes on the summary screen and paste in an address from the same wallet):

Screenshot 2021-04-17 at 19 40 51

**Comment by nikolaglumac** [comment]
@darko-mijic did you approve this text copy?

**Comment by aleksandardjordjeviciohk** [comment]
> @aleksandardjordjeviciohk please also prevent autofocus of the token amount field in case user enters the same address (for this you need to click on "SEND" in one of the token boxes on the summary screen and paste in an address from the same wallet):
> 
> Screenshot 2021-04-17 at 19 40 51

Fixed

**Review by ManusMcCole — APPROVED** [review]
LGTM 👍

**Comment by gabriela-ponce** [comment]
@aleksandardjordjeviciohk The issues for the "Receiver" input look fixed to me and now I can see the tooltip on hover.
I tested all the inputs in 17660 and I found two minor inconsistencies, that I am not sure if they should be addressed on this PR, let me know what you think:

- The "Recommended" placeholder on the smash server dropdown is highlighted. I can't reproduce this in develop or production. Check [this screenshot](https://drive.google.com/file/d/1d9MlgM9Dj_1etgZVpl3dFMxY9tJ8faCQ/view?usp=sharing).
- The inputs for Smash server URL and for wallet's name don't show the tooltip on hover, only when focusing on the input. This reproduces also in production, just wondering if it should be consistent across all inputs (as now they show the error on hover). Check videos [1](https://drive.google.com/file/d/1dXQ9zsDqOvDlE6JYz7whjq0xiNcna07e/view?usp=sharing) and [2](https://drive.google.com/file/d/1-WZJkAg086KfXl1lvICpNdhumzvFZjfy/view?usp=sharing).

**Comment by aleksandardjordjeviciohk** [comment]
> @aleksandardjordjeviciohk The issues for the "Receiver" input look fixed to me and now I can see the tooltip on hover.
> I tested all the inputs in 17660 and I found two minor inconsistencies, that I am not sure if they should be addressed on this PR, let me know what you think:
> 
> * The "Recommended" placeholder on the smash server dropdown is highlighted. I can't reproduce this in develop or production. Check [this screenshot](https://drive.google.com/file/d/1d9MlgM9Dj_1etgZVpl3dFMxY9tJ8faCQ/view?usp=sharing).
> * The inputs for Smash server URL and for wallet's name don't show the tooltip on hover, only when focusing on the input. This reproduces also in production, just wondering if it should be consistent across all inputs (as now they show the error on hover). Check videos [1](https://drive.google.com/file/d/1dXQ9zsDqOvDlE6JYz7whjq0xiNcna07e/view?usp=sharing) and [2](https://drive.google.com/file/d/1-WZJkAg086KfXl1lvICpNdhumzvFZjfy/view?usp=sharing).

@gabriela-ponce pls recheck, both of these issues are fixed now.

**Comment by gabriela-ponce** [comment]
@aleksandardjordjeviciohk Fixes look good, just a small issue on "Smash server" and "Wallet name" inputs:

- In the latest build, the tooltip is not displayed when the input is on focus, [it's only shown on hover](https://drive.google.com/file/d/1K9kYmNaAydpDDm2u0NgSgZFJh_C9Oven/view?usp=sharing). To be consistent with the rest of the inputs, it should be displayed [on both instances.](https://drive.google.com/file/d/1PLPkzUyROIGvFVVHUnX8yAEPLilqX6fU/view?usp=sharing)

**Comment by aleksandardjordjeviciohk** [comment]
> @aleksandardjordjeviciohk Fixes look good, just a small issue on "Smash server" and "Wallet name" inputs:
> 
> * In the latest build, the tooltip is not displayed when the input is on focus, [it's only shown on hover](https://drive.google.com/file/d/1K9kYmNaAydpDDm2u0NgSgZFJh_C9Oven/view?usp=sharing). To be consistent with the rest of the inputs, it should be displayed [on both instances.](https://drive.google.com/file/d/1PLPkzUyROIGvFVVHUnX8yAEPLilqX6fU/view?usp=sharing)

@DominikGuzei @gabriela-ponce it appears that InlineEditingInput which is using Input component is actually not handling hover/focus for tooltip in the correct way. I have removed custom part of the implementation from Daedalus which overwrites this part, but still we have a false case scenario with hover.

**Comment by aleksandardjordjeviciohk** [comment]
> > @aleksandardjordjeviciohk Fixes look good, just a small issue on "Smash server" and "Wallet name" inputs:
> > 
> > * In the latest build, the tooltip is not displayed when the input is on focus, [it's only shown on hover](https://drive.google.com/file/d/1K9kYmNaAydpDDm2u0NgSgZFJh_C9Oven/view?usp=sharing). To be consistent with the rest of the inputs, it should be displayed [on both instances.](https://drive.google.com/file/d/1PLPkzUyROIGvFVVHUnX8yAEPLilqX6fU/view?usp=sharing)
> 
> @DominikGuzei @gabriela-ponce it appears that InlineEditingInput which is using Input component is actually not handling hover/focus for tooltip in the correct way. I have removed custom part of the implementation from Daedalus which overwrites this part, but still we have a false case scenario with hover.

I found the solution which covers this case, now all the Input fields handle tooltips in the same way.

**Review by gabriela-ponce — APPROVED** [review]
Looks good, nice job 🎉

**Review by nikolaglumac — CHANGES_REQUESTED** [review]
@aleksandardjordjeviciohk can you please check why this new tooltip is shown as an error one (in red instead of orange) in the latest builds?

Screenshot 2021-04-23 at 08 16 01
Screenshot 2021-04-23 at 08 16 28

**Comment by aleksandardjordjeviciohk** [comment]
> @aleksandardjordjeviciohk can you please check why this new tooltip is shown as an error one (in red instead of orange) in the latest builds?
> 
> Screenshot 2021-04-23 at 08 16 01
> Screenshot 2021-04-23 at 08 16 28

Updated with new version of React Polymorph.

Screen Shot 2021-04-26 at 12 47 41 PM

Screen Shot 2021-04-26 at 12 47 56 PM

File: source/renderer/app/components/wallet/WalletSendForm.js [hunk 7]
```diff
@@ -628,7 +635,9 @@ export default class WalletSendForm extends Component<Props, State> {
         ];
       },
     ]);
-    this.form.$(newAsset).focus();
+    if (!this.isAddressFromSameWallet()) {
```
Comment (nikolaglumac) [inline source/renderer/app/components/wallet/WalletSendForm.js:638]: @aleksandardjordjeviciohk this should be reverted.

Comment (aleksandardjordjeviciohk) [reply source/renderer/app/components/wallet/WalletSendForm.js]: Updated

File: source/renderer/app/components/wallet/WalletSendForm.js [hunk 8]
```diff
@@ -333,6 +341,11 @@ export default class WalletSendForm extends Component<Props, State> {
               if (isValid && isAdaAmountValid) {
                 this.calculateTransactionFee();
               } else {
+                const { selectedAsset } = this.props;
```
Comment (nikolaglumac) [inline source/renderer/app/components/wallet/WalletSendForm.js:344]: @aleksandardjordjeviciohk please remove this part.

File: source/renderer/app/components/wallet/WalletSendForm.js [hunk 9]
```diff
@@ -515,6 +528,12 @@ export default class WalletSendForm extends Component<Props, State> {
     }
   };
 
+  focusAssetField = (fingerprint: string) => {
```
Comment (nikolaglumac) [inline source/renderer/app/components/wallet/WalletSendForm.js:531]: @aleksandardjordjeviciohk please remove this method.

File: source/renderer/app/components/wallet/WalletSendForm.js [hunk 10]
```diff
@@ -825,6 +876,11 @@ export default class WalletSendForm extends Component<Props, State> {
                       handleSubmitOnEnter={this.handleSubmitOnEnter}
                       clearAssetFieldValue={this.clearAssetFieldValue}
                       onChangeAsset={this.onChangeAsset}
+                      autoFocus={
```
Comment (nikolaglumac) [inline source/renderer/app/components/wallet/WalletSendForm.js:879]: @aleksandardjordjeviciohk please remove this part.

File: source/renderer/app/components/wallet/WalletSendForm.js [hunk 11]
```diff
@@ -299,8 +300,14 @@ export default class WalletSendForm extends Component<Props, State> {
     return receiverField.value.length > 0;
   };
 
-  hasAssetValue = (asset: Field) => {
-    return get(asset, 'value', false);
+  isAddressFromSameWallet = () => {
+    const { isAddressFromSameWallet } = this.props;
+    const receiverField = this.form.$('receiver');
+    return (
+      this.hasReceiverValue() &&
```
Comment (nikolaglumac) [inline source/renderer/app/components/wallet/WalletSendForm.js:307]: @aleksandardjordjeviciohk this part is repeated in two places - can we extract it into a method and reuse it in both?

File: source/renderer/app/components/wallet/WalletSendForm.js [hunk 12]
```diff
@@ -700,27 +722,51 @@ export default class WalletSendForm extends Component<Props, State> {
       !this.hasAvailableAssets ? styles.disabled : null,
       'primary',
     ]);
+
+    const receiverFieldClasses = classNames([
+      styles.receiverInput,
+      this.hasReceiverValue() &&
```
Comment (nikolaglumac) [inline source/renderer/app/components/wallet/WalletSendForm.js:728]: @aleksandardjordjeviciohk this part is repeated in two places - can we extract it into a method and reuse it in both?

File: source/renderer/app/components/wallet/WalletSendForm.js [hunk 13]
```diff
@@ -731,6 +777,7 @@ export default class WalletSendForm extends Component<Props, State> {
                 <button
                   onClick={() => this.handleOnReset()}
                   className={styles.clearReceiverButton}
+                  tabIndex={-1}
```
Comment (nikolaglumac) [inline source/renderer/app/components/wallet/WalletSendForm.js:780]: @aleksandardjordjeviciohk why do we need this? What is it for?

File: source/renderer/app/components/wallet/WalletSendForm.js [hunk 14]
```diff
@@ -330,6 +337,17 @@ export default class WalletSendForm extends Component<Props, State> {
...
               const isAdaAmountValid = adaAmountField.isValid;
+              const { selectedAsset } = this.props;
+              const shouldFocus = !this.isAddressFromSameWallet();
+              if (selectedAsset && !shouldFocus) {
+                const newAsset = `asset_${selectedAsset.fingerprint}`;
+                const newAssetField = this.form.$(newAsset);
+                const { name: fieldName } = newAssetField;
+                const focusableField = this.focusableFields[fieldName];
+                if (focusableField && focusableField.inputElement) {
+                  focusableField.inputElement.current.blur();
```
Comment (nikolaglumac) [inline source/renderer/app/components/wallet/WalletSendForm.js:348]: @aleksandardjordjeviciohk this feel quite hacky...

File: source/renderer/app/components/wallet/WalletSendForm.js (hunk 11 above)
Comment (aleksandardjordjeviciohk) [reply source/renderer/app/components/wallet/WalletSendForm.js:307]: Fixed

File: source/renderer/app/components/wallet/WalletSendForm.js (hunk 12 above)
Comment (aleksandardjordjeviciohk) [reply source/renderer/app/components/wallet/WalletSendForm.js:728]: Fixed

File: source/renderer/app/components/wallet/WalletSendForm.js (hunk 13 above)
Comment (aleksandardjordjeviciohk) [reply source/renderer/app/components/wallet/WalletSendForm.js:780]: There was an issue with tab indexing and this button was taking focus out of tab focus on input fields, so we are now skipping it.

File: source/renderer/app/components/wallet/WalletSendForm.js (hunk 14 above)
Comment (aleksandardjordjeviciohk) [reply source/renderer/app/components/wallet/WalletSendForm.js:348]: Updated

File: package.json [hunk 15]
```diff
@@ -234,7 +234,7 @@
     "react-intl": "2.7.2",
     "react-lottie": "1.2.3",
     "react-markdown": "4.3.1",
-    "react-polymorph": "0.9.8-rc.21",
+    "react-polymorph": "0.9.7-next.1",
```
Comment (nikolaglumac) [inline package.json:237]: @aleksandardjordjeviciohk is this the latest version of RP?

File: source/renderer/app/components/widgets/forms/InlineEditingInput.js [hunk 16]
```diff
@@ -301,11 +301,9 @@ export default class InlineEditingInput extends Component<Props, State> {
           type="text"
           maxLength={maxLength}
           label={label}
-          onFocus={this.onFocus}
```
Comment (nikolaglumac) [inline source/renderer/app/components/widgets/forms/InlineEditingInput.js:304]: @aleksandardjordjeviciohk please revert this

File: source/renderer/app/components/widgets/forms/InlineEditingInput.js [hunk 17]
```diff
@@ -301,11 +301,9 @@ export default class InlineEditingInput extends Component<Props, State> {
...
-          onBlur={this.onBlur}
```
Comment (nikolaglumac) [inline source/renderer/app/components/widgets/forms/InlineEditingInput.js:305]: @aleksandardjordjeviciohk please revert this

File: source/renderer/app/components/widgets/forms/InlineEditingInput.js [hunk 18]
```diff
@@ -301,11 +301,9 @@ export default class InlineEditingInput extends Component<Props, State> {
...
           onChange={this.onChange}
           onKeyDown={(event) => this.handleInputKeyDown(event)}
-          error={isActive ? error : !!error}
```
Comment (nikolaglumac) [inline source/renderer/app/components/widgets/forms/InlineEditingInput.js:308]: @aleksandardjordjeviciohk please check the impact of this.

File: source/renderer/app/components/widgets/forms/InlineEditingInput.js (hunk 16 above)
Comment (aleksandardjordjeviciohk) [reply source/renderer/app/components/widgets/forms/InlineEditingInput.js:304]: Done

File: source/renderer/app/components/widgets/forms/InlineEditingInput.js (hunk 17 above)
Comment (aleksandardjordjeviciohk) [reply source/renderer/app/components/widgets/forms/InlineEditingInput.js:305]: Done

File: source/renderer/app/components/widgets/forms/InlineEditingInput.js (hunk 18 above)
Comment (aleksandardjordjeviciohk) [reply source/renderer/app/components/widgets/forms/InlineEditingInput.js:308]: Done

**Comment by gabriela-ponce** [comment]
@aleksandardjordjeviciohk I checked the behavior for all inputs and most of them are ok, but found a minor inconsistency:

- For the "Smash server URL" the tooltip is shown only on focus at first. Then, if you insert an incorrect URL and submit, all the tooltips would be displayed, even if the input is not in focus or hovered. Check [this video.](https://drive.google.com/file/d/1J9b4dj7kAT-TeKcj8A-BXPZtW_nhOvAb/view?usp=sharing)

Also, I have a few blockers to complete the testing of this task:

- On build 17794, there are no installers for macOS.
- Voting is not enabled on Testnet to check the corresponding inputs.

**Comment by nikolaglumac** [comment]
> On build 17794, there are no installers for macOS.

They will show up eventually. Please stand by 🙏 

> Voting is not enabled on Testnet to check the corresponding inputs.

You can skip that part as it is not easy to enable it at the moment...

**Comment by DominikGuzei** [comment]
@input-output-hk/daedalus-review I updated this PR to point to the `1.0.0-rc.1` release candidate 🎉

**Comment by nikolaglumac** [comment]
> For the "Smash server URL" the tooltip is shown only on focus at first. Then, if you insert an incorrect URL and submit, all the tooltips would be displayed, even if the input is not in focus or hovered. Check this video.

@gabriela-ponce this is the same behavior as in production or develop builds. The input remains focused (even if you click outside of it) after you submit an incorrect URL.

**Review by nikolaglumac — APPROVED** [review]
Works good!

**Pull request merged**